client/rpc-servers/src/lib.rs

```diff
@@ -108,125 +72,106 @@ impl ServerMetrics {
-	server_metrics: ServerMetrics,
-	tokio_handle: tokio::runtime::Handle,
-) -> io::Result<ws::Server> {
-	let max_payload = maybe_max_payload_mb
+	module: RpcModule<M>,
+	rt: tokio::runtime::Handle,
+) -> Result<WsServerHandle, anyhow::Error> {
+	let max_request_body_size = maybe_max_payload_mb
 		.map(|mb| mb.saturating_mul(MEGABYTE))
 		.unwrap_or(RPC_MAX_PAYLOAD_DEFAULT);
-	let max_out_buffer_capacity = maybe_max_out_buffer_capacity_mb
-		.map(|mb| mb.saturating_mul(MEGABYTE))
-		.unwrap_or(WS_MAX_BUFFER_CAPACITY_DEFAULT);
-
-	if max_payload > max_out_buffer_capacity {
-		log::warn!(
-			"maximum payload ({}) is more than maximum output buffer ({}) size in ws server, the payload will actually be limited by the buffer size",
-			max_payload,
-			max_out_buffer_capacity,
-		)
+	let max_connections = max_connections.unwrap_or(WS_MAX_CONNECTIONS);
+
+	let mut builder = WsServerBuilder::default()
+		.max_request_body_size(max_request_body_size as u32)
+		.max_connections(max_connections as u64)
+		.custom_tokio_runtime(rt.clone());
+
+	log::info!("Starting JSONRPC WS server: addr={}, allowed origins={:?}", addr, cors);
+
+	if let Some(cors) = cors {
+		// Whitelist listening address.
+		builder = builder.set_allowed_hosts([
+			format!("localhost:{}", addr.port()),
+			format!("127.0.0.1:{}", addr.port()),
+		])?;
+
+		// Set allowed origins.
+		builder = builder.set_allowed_origins(cors)?;
 	}
 
-	ws::ServerBuilder::with_meta_extractor(io, |context: &ws::RequestContext| {
-		context.sender().into()
-	})
-	.event_loop_executor(tokio_handle)
-	.max_payload(max_payload)
-	.max_connections(max_connections.unwrap_or(WS_MAX_CONNECTIONS))
-	.max_out_buffer_capacity(max_out_buffer_capacity)
-	.allowed_origins(map_cors(cors))
-	.allowed_hosts(hosts_filtering(cors.is_some()))
-	.session_stats(server_metrics)
-	.start(addr)
-	.map_err(|err| match err {
-		ws::Error::Io(io) => io,
-		ws::Error::ConnectionClosed => io::ErrorKind::BrokenPipe.into(),
-		e => {
-			error!("{}", e);
-			io::ErrorKind::Other.into()
-		},
-	})
-}
+	let server = tokio::task::block_in_place(|| rt.block_on(builder.build(addr)))?;
+
+	let rpc_api = build_rpc_api(module);
+	let handle = server.start(rpc_api)?;
 
-fn map_cors<T: for<'a> From<&'a str>>(cors: Option<&Vec<String>>) -> http::DomainsValidation<T> {
-	cors.map(|x| x.iter().map(AsRef::as_ref).map(Into::into).collect::<Vec<_>>())
-		.into()
+	Ok(handle)
 }
 
-fn hosts_filtering(enable: bool) -> http::DomainsValidation<http::Host> {
-	if enable {
-		// NOTE The listening address is whitelisted by default.
-		// Setting an empty vector here enables the validation
-		// and allows only the listening address.
-		http::DomainsValidation::AllowOnly(vec![])
-	} else {
-		http::DomainsValidation::Disabled
-	}
+fn build_rpc_api<M: Send + Sync + 'static>(mut rpc_api: RpcModule<M>) -> RpcModule<M> {
+	let mut available_methods = rpc_api.method_names().collect::<Vec<_>>();
+	// NOTE(niklasad1): substrate master doesn't have this.
```

Comment:
can you elaborate? Is this not useful?

Reply:
Yeah, it useful it's just that I added `rpc_methods` to the list of available methods.

Furthermore, before (substrate master) `"rpc_methods"` returned all methods except `"rpc_methods"` but I realize that it's redundant to I think we could have the same behavior and remove `available_methods.push("rpc_methods");` but doesn't really matter to me.

Reply:
The note was bad :)

Reply:
Okay, yeah, sorry, my second question should be: "Is this note useful", cause my main point was that I didn't understand what was it about.